docs/en/tutorials/getting-started.rst

```diff
@@ -233,33 +233,154 @@ entity definition:
+
+Adding the Entity behaviors
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You have two options to define the accessors mutators that will
+update the state of your objects.
+
+**Getters and setters**
+
+The most popular method is to create getter and setter methods
+for each property (except ``$id``). The use of mutators allows
+Doctrine to hook into calls which manipulate the entities in ways
+that it could not if you just directly set the values with ``entity#field = foo;``.
+
+The id field has no setter since, generally speaking, your code
+should not set this value since it represents a database id value.
+(Note that Doctrine itself can still set the value using the
+Reflection API instead of a defined setter function)
+
+.. note::
+
+    This method, although very common, is **not** the best approach if
+    you want a perfect domain design. It can lead to unexpected behavior
+    when dealing with different steps of logic in between different
+    calls to ``EntityManager::flush()``, for example when you change
+    the state of an object via event listeners or lifecycle callbacks.
+
+**Mutators and DTOs**
+
+We recommend using a rich model design and rely on more complex
+mutators, ideally based on DTOs.
+This way, you should **not** use getters nor setters, and instead,
+implement methods that represent the **behavior** of your domain.
+
+For example, when having a ``User`` entity, we could foresee
+the following kind of optimization.
+
+Before, an anemic model with getters and setters:
+
+.. configuration-block::
+
+    .. code-block:: php
+
+        <?php
+        class User {
+            private $username;
+            private $passwordHash;
+            private $bans;
+
+            public function getUsername() :string {
+                return $this->username;
+            }
+
+            public function setUsername(string $username) {
+                $this->username = $username;
+            }
+
+            public function getPasswordHash() : string {
+                return $this->passwordHash;
+            }
+
+            public function setPasswordHash(string $passwordHash) {
+                $this->passwordHash = $passwordHash;
+            }
+
+            public function addBan(Ban $ban) {
+                $this->bans[] = $ban;
+            }
         }
 
-        public function getName()
+After, a rich model with proper accessors and mutators:
+
+.. configuration-block::
+
+    .. code-block:: php
+        <?php
+        class User
         {
-            return $this->name;
+            private $banned;
+            private $username;
+            private $passwordHash;
+            private $bans;
+
+            public function toNickname() : string
+            {
+                return $this->username;
+            }
+
+            public function authenticate(string $password, callable $checkHash) : bool
+            {
+                return $checkHash($password, $this->passwordHash) && ! $this->hasActiveBans();
```

Comment:
what about cases with "password_needs_rehash"?

Reply:
It's an example retrieved from Marco's talk about doctrine best practices, but of course it can be changed :)

Reply:
@fesor I think this is a good example of how a rich domain model aids in better maintainability later on. Not sure if it would fit a simplified example though. Maybe just skip it for now?

Reply:
What I thought is... `password_hash` API is already a good abstraction... maybe we don't need to pass closure for that specific case? 

@Ocramius what is your thoughts about `password_needs_rehash` case?

Reply:
Yes, probably this something that could be covered in cookbook (like how to deal with dependencies).

Reply:
Hi guys.
I appreciate the effort you put in updating the documentation and promote better programming practices.
I agree with @fesor that the "dealing with dependencies" topic must be covered. Yes, it can be in some cookbook, but it must be covered because it will be an issue people will run into very quickly and not finding anything related to this will probably make them to give up and go to the well known "don't put behavior in your entities, just put data and make some services that will operate on your entities through some setters".